src/constraint/pip.rs

```diff
@@ -0,0 +1,670 @@
+//!
+use std::cmp::Ordering;
+
+use derive_more::Display;
+use nom::{
+    IResult, Parser,
+    branch::alt,
+    bytes::complete::{tag, take_while1},
+    character::complete::{char, digit1, multispace0, u32},
+    combinator::{eof, map_res, opt, value},
+    multi::{many1, separated_list1},
+    sequence::{delimited, pair, preceded, terminated},
+};
+use thiserror::Error;
+use tracing::warn;
+
+use super::{Constraint, fallback};
+use crate::{Fetcher, Revision};
+
+/// A version in pip.
+/// See [Version Specifiers](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers) for more information.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct PipVersion {
+    /// The epoch of the version; most versions have an epoch of 0, which is the default.
+    /// After, for example, changing from a `yyyy.mm.dd` version format to a semver format, a project might increment its epoch to capture the change.
+    /// That would allow the project to indicate that every semver version was later than any date-marked version.
+    /// In a version `0!1.2.3.rc4.post5.dev6`, the epoch is 0.
+    epoch: u32,
+
+    /// The release segments of the version.
+    /// These are what we most commonly picture when we imagine a version.
+    /// For example, in the version `1.2.3`, the release segments are `[1, 2, 3]`.
+    /// In a version `0!1.2.3.rc4.post5.dev6`, the release segments are `[1, 2, 3]`.
+    release_segments: Vec<u32>,
+
+    /// An optional prerelease.
+    /// In a version `0!1.2.3.rc4.post5.dev6`, the prerelease is `rc 4`.
+    /// In a version `0!1.2.3.pre4.post5.dev6`, the prerelease is `rc 4`.
+    /// In a version `0!1.2.3.prerelease4.post5.dev6`, the prerelease is `rc 4`.
+    /// In a version `0!1.2.3.a4.post5.dev6`, the prerelease is `a 4`.
+    /// In a version `0!1.2.3.b4.post5.dev6`, the prerelease is `b 4`.
+    /// In a version `0!1.2.3.c4.post5.dev6`, the prerelease is `rc 4`.
+    pre_release: Option<PreRelease>,
+
+    /// An optional sub-patch post-release revision.
+    /// In a version `0!1.2.3.rc4.post5.dev6`, the post release is `5`.
+    /// In a version `0!1.2.3.rc4.r5.dev6`, the post release is `5`.
+    /// In a version `0!1.2.3.rc4.rev5.dev6`, the post release is `5`.
+    post_release: Option<u32>,
+
+    /// An optional dev-release, 0 by default.
+    /// Dev releases are newer than the post-releases of the prior version and older than pre-releases of their version.
+    /// In a version `0!1.2.3.rc4.post5.dev6`, the dev release is `6`.
+    dev_release: Option<u32>,
+}
+
+#[derive(Error, Clone, PartialEq, Eq, Debug, Display)]
+pub enum PipConstraintError {
+    #[display("ConstraintParseError({constraints}, {message})")]
+    ConstraintParseError {
+        constraints: String,
+        message: String,
+    },
+
+    #[display("VersionParseError({version}, {message})")]
+    VersionParseError { version: String, message: String },
+}
```

Comment:
[nit] you already know these are errors
[nit] error messages ideally read more like english
[nit] you're deriving `Error` with `thiserror` but using `derive_more::Display`, prefer to pick one or the other (`thiserror` or `derive_more`). the suggestion below assumes `thiserror`.
```suggestion
pub enum PipConstraintError {
    #[error("parse constraints {constraints:?}: {message:?})")]
    ConstraintParse {
        constraints: String,
        message: String,
    },

    #[error("parse version {version:?}: {message:?})")]
    VersionParse { version: String, message: String },
}
```